docs/dev/swagger.yaml

```diff
@@ -0,0 +1,1075 @@
+        type: string
+      salutation:
+        type: string
+      suffix:
+        type: string
+      affiliation:
+        $ref: '#/definitions/LocaleObject'
+      country:
+        type: string
+      url:
+        type: string
+      email:
+        type: string
+      orcid:
+        type: string
+      groups:
+        type: array
+        items:
+          $ref: '#/definitions/UserGroup'
+      biography:
+        $ref: '#/definitions/LocaleObject'
+      signature:
+        $ref: '#/definitions/LocaleObject'
+      authId:
+        type: integer
+      authString:
+        type: string
+      phone:
+        type: string
+      mailingAddress:
+        type: string
+      billingAddress:
+        type: string
+      interests:
+        type: array
+        items:
+          $ref: '#/definitions/UserInterest'
+      gossip:
+        $ref: '#/definitions/LocaleObject'
+      disabled:
+        type: boolean
+      disabledReason:
+        type: string
+      dateRegistered:
+        type: string
+      dateValidated:
+        type: string
+      dateLastLogin:
+        type: string
+      mustChangePassword:
+        type: boolean
+  ReviewerSummary:
+    type: object
+    properties:
+      id:
+        type: integer
+      _href:
+        type: string
+      userName:
+        type: string
+      fullName:
+        type: string
+      affiliation:
+        $ref: '#/definitions/LocaleObject'
+      groups:
+        type: array
+        items:
+          $ref: '#/definitions/UserGroup'
+      interests:
+        type: array
+        items:
+          $ref: '#/definitions/UserInterest'
+      gossip:
+        $ref: '#/definitions/LocaleObject'
+      reviewsActive:
+        type: integer
+      reviewsCompleted:
+        type: integer
+      averageReviewCompletionDays:
+        type: integer
+      dateLastReviewAssignment:
+        type: string
+      disabled:
+        type: boolean
+  UserGroup:
+    type: object
+    properties:
+      id:
+        type: integer
+      name:
+        $ref: '#/definitions/LocaleObject'
+      abbrev:
+        $ref: '#/definitions/LocaleObject'
+      roleId:
+        type: integer
+      showTitle:
+        type: boolean
+      permitSelfRegistration:
+        type: boolean
+      recommendOnly:
+        type: boolean
+  UserInterest:
+    type: object
+    properties:
+      id:
+        type: integer
+      interest:
+        type: string
+  LocaleObject:
+    type: object
+    description: Key/value pairs containing strings by locale key.
+    example:
+      en_US: Filename
+      fr_CA: Nom de fichier
+  ErrorResponse:
+    type: object
+    description: A response object that will be returned when the request could not be fulfilled.
+    properties:
+      error:
+        type: string
+      errorMessage:
+        type: string
+
+paths:
+  '/submissions':
+    get:
+      tags:
+        - Submissions
+      summary: Get a list of submissions.
+      description: |
+        **OJS 3.1 Target**:
```

Comment:
Still refers to OJS. (Is it possible to generalize and move some of this description into `pkp-lib` rather than copying/replacing?)

Reply:
I have done some research. it's possible to split a swagger file into multiple documents. Where would be a suitable location in `pkp-lib`? Can I store shared definitions in `docs/dev/swagger` folder ?

Reply:
so I have done the initial work of moving shared `swagger.json` definitions to `pkp-lib`. The problem at the moment is that I have no way to validate the syntax. 

I came across couple of nodejs scripts (https://github.com/mohsen1/multi-file-swagger-example & https://github.com/dzuluaga/json-refs-test) to do so but I was unable to successfully test the swagger document. 

Basically instructions like `$ref: '../../lib/pkp/docs/dev/swagger/produces.yaml'` are supposed to be replaced by the target file content. However when I try it, it's like if the file cannot be resolved.

As per this comment,  `options.relativeBase` needs to be set (https://github.com/whitlockjc/json-refs/issues/78#issuecomment-187930327). It seems that `options.relativeBase` is supported deprecated in general and we should use  `options.location` now (https://github.com/whitlockjc/json-refs/issues/76#issuecomment-301524186)

@NateWr since you are more familiar with NodeJS, any hint ? Thanks.

Reply:
I've never worked with `json-refs` so I'm not sure. Have you tried the `options.location` approach? No luck? 

When are you trying to validate the syntax? Is this when the tests are run? If you are just looking to validate a JSON file, it might make more sense to lint it? We do that with webpack for the JS in the UI Library, so maybe you could piggy-back on that?

Reply:
I noticed that the online Swagger Editor had a "Convert to JSON" feature. So I dumped your OMP swagger file with it's `$ref`s in, and converted it, and I got refs that look like this:

```
  "definitions": {
    "$ref": "../../lib/pkp/docs/dev/swagger/sharedDefinitions.yaml",
    "Submission": {
      "type": "object",
      "properties": {
        "$ref": "../../lib/pkp/docs/dev/swagger/sharedSubmissionProperties.yaml",
        "chapters": {
          "type": "array",
          "items": {
            "$ref": "#/definitions/Chapter"
          }
        }
      }
    },
```

I know in JSON you have to use `"` not `'`, and maybe because it starts with `$` you need to quote the property too. Could that be the issue you're facing?

Reply:
> Have you tried the options.location approach? No luck?

Yes. But it didn't work.

Alec suggested to move some of the shared swagger spec to pkp-lib. I did some research and saw that it was possible to split a swagger spec into multiple documents via `$ref`. 

After splitting `swagger.yaml` into multiple files, I needed to test the correctness of the spec. We also need to find a way to generate a complete swagger spec (replace refs with their target content) so that a user willing to test from Swagger editor can generate a full swagger spec in a single document. 

Thus instead of having 

> ...
"properties": {
  "$ref": "../../lib/pkp/docs/dev/swagger/sharedSubmissionProperties.yaml",
  "chapters": {
...

You would have 

> ...
"properties": {
  id:
  type: integer
_href:
  type: string
title:
  $ref: '#/definitions/LocaleObject',
  "chapters": {
...


ps: When I move all referenced specs into the same folder or subfolder and run let's say `multi-file-swagger swagger.yaml > index.json` everything works fine, meaning all `$refs` are replaced. But when I move the files to `lib/pkp/docs/dev/swagger`, no luck.

Reply:
If you have the swagger file in JSON format, you might be able to just use regular node ecosystem tools. This [json-refs](https://www.npmjs.com/package/json-refs) package looks like it has a [command-line tool](https://github.com/whitlockjc/json-refs/blob/master/docs/CLI.md) which might compile a JSON file using it's refs. If that works, we could add it to our build scripts and commit the compiled files to version control.

Reply:
yes that's the plan but I need to figure out the conversion first :sweat:

Reply:
Sorry, maybe I'm misunderstanding something, but do you mean converting YAML to JSON? The online [Swagger Editor](http://editor.swagger.io/) can do that from our existing all-in-one YAML file. In the menu go to File > Convert and save as JSON.

Reply:
Sorry, let me try to explain again. 

We had a single `swagger.yaml` containing all the specs. Now we are moving shared specs into other remote files : `definitions.yaml`, `Submission.yaml` and `SubmissionSummary.yaml`, etc. 

Let's say a user would like to load OJS REST API swagger specs into Swagger editor. At the moment he could load `swagger.yaml` into the online swagger editor because it's a single file containing all the specs. Once we split into multiple files, `swagger.yaml` will contain references to other files (like `definitions.yaml`, `Submission.yaml` and `SubmissionSummary.yaml`) and Swagger editor will not be able to load load these remote spec because these files are located in `pkp-lib` which the online editor do not have access to. 


1. Like you mentioned above, it would be nice to have a `swagger.json` file automatically generated by converting `swagger.yaml`. That's what I am having trouble with. 

2. At the moment I also need to test that the changes I just made (breaking `swagger.yaml` into multiple files) is correct. And to do so I am trying to generate a single file `swagger.yaml` or `swagger.json` from (`swagger.yaml`, `definitions.yaml`, `Submission.yaml` and `SubmissionSummary.yaml`, etc. ). I found a couple of NodeJS tools (see above) that I tried. But instead of replacing `$ref`s with the content of the remote files I still see the paths.

I hope you have a better understanding of the issue now.

Reply:
Ah, I was thinking that we would abandon YAML and move our specs over to JSON. So I'd see the workflow being:

1. Convert YAML files to JSON files and commit those to version control.
2. When we want a single file from these JSON files, use the CLI tool from the `json-refs` package to generate it based on the JSON.

So instead of trying to use a NodeJS tool to read the YAML files, we just go all-in with JSON. Would that work?

Reply:
Okay with replacing YAML with JSON. but

> When we want a single file from these JSON files, use the CLI tool from the json-refs package to generate it based on the JSON.

That's the issue I am having at the moment. I am unable to create a single file out of these JSON files. see https://github.com/pkp/omp/pull/541#discussion_r191517617

Reply:
Have you tried the npm package I recommended (`json-refs`)? You mentioned trying two nodejs scripts that were specific to swagger, but I don't think we need anything specific to swagger if we have the JSON files ready.

Reply:
I think I have done so.  I converted `swagger.yaml` to `swagger.json` and then `npm resolve swagger.json`.

Reply:
Ah ok. Do you have the code with the JSON file somewhere I could try it out?

Reply:
I think you need to run `json-refs` from the parent directory. So it would be:

```
json-refs food/bar/test.json
```

I found that worked for me when reproducing the error you describe.

(I'm thinking about that disappearing `info` prop. Do `json-refs` actually support property merging? If not, that could be a problem...)

Reply:
it looks like `$ref` must stand alone and should not be used for properties. 
(see https://stackoverflow.com/a/31011359)